File: modules/x-engine-jsi-legacy/iOS/Class/router/JSIOldRouterModule.m
//
//  JSIVueRouterModule.h
//  ModuleApp
//
//  Created by zk on 2021/3/14.
//  Copyright © 2021 zkty-team. All rights reserved.
//

#import "JSIOldRouterModule.h"
#import "JSIContext.h"
#import "iDirectManager.h"
#import "NativeContext.h"
#import "NSURL+QueryDictionary.h"

@interface JSIOldRouterModule ()
@property (nonatomic, strong)   id<iDirectManager>  directors;
@end
@implementation JSIOldRouterModule
JSI_MODULE(JSIOldRouterModule)

- (NSString*) moduleId{
    return @"com.zkty.module.router";
}

-(void)afterAllJSIModuleInited {
    self.directors = [[NativeContext sharedInstance] getModuleByProtocol:@protocol(iDirectManager)];
}


static NSString *const kQueryBegin          = @"?";
static NSString *const kFragmentBegin       = @"#";
static NSString *const kSlash               = @"/";

+ (NSString*)SPAUrl2StandardUrl:(NSString*)raw {
    int questionMark = -1;
    int hashtagMark = -1;

    for (int i=0;i< raw.length;i++){
        char cc= [raw characterAtIndex:i];

        if(cc == '#' && hashtagMark == -1){
            hashtagMark=i;
        }
        else if(cc == '?' && questionMark == -1){
            questionMark=i;
        }
    }
    if(questionMark != -1 && hashtagMark != -1){
        NSString* sub1= [raw substringToIndex:hashtagMark];
        NSString* sub2= [raw substringWithRange:NSMakeRange(hashtagMark, questionMark-hashtagMark)];
        NSString* sub3=[ [raw substringFromIndex:questionMark]stringByAddingPercentEncodingWithAllowedCharacters:[NSCharacterSet URLQueryAllowedCharacterSet]];
        return [NSString stringWithFormat:@"%@%@%@",sub1,sub3,sub2] ;
    }
    return raw;
}
// 区分 vue 的 url
// 与标准的 rfc url
+ (NSDictionary*) convertRouter2JSIModel:(NSDictionary*) dict{
    NSMutableDictionary* ret= [NSMutableDictionary new];
    ret[@"scheme"] =dict[@"type"];
    
    NSString* host= dict[@"uri"];
    NSString* url_need_check = @"";
    
    if([dict[@"type"] isEqualToString:@"microapp"]){
        // 一定是非标 url
        // microapp id，应该组装成 http://, 以供 NSURL 标准化
        url_need_check = [NSString stringWithFormat:@"http://%@#%@",host, dict[@"path"]];
    }
    else{
        // 其他的直接拼接
        // 可能是普通 url
        // 也可能是非标 url
        if([host containsString:@"#"])
            url_need_check = [NSString stringWithFormat:@"%@%@",host, dict[@"path"]];
        else
            url_need_check = [NSString stringWithFormat:@"%@#%@",host, dict[@"path"]];

    }
    // 全部转化为标准 url
    NSString* standard_url = [JSIOldRouterModule SPAUrl2StandardUrl:url_need_check];

    NSURL * url = [NSURL URLWithString:standard_url];
 
    if(url.port){
        ret[@"host"]=[NSString stringWithFormat:@"%@:%@" , url.host, url.port];
    }else{
        ret[@"host"]=url.host;
    }
    // https://tools.ietf.org/html/rfc3986#page-22
    if(url.path && url.path.length != 0 )
        ret[@"pathname"]=url.path;
    else
        ret[@"pathname"]=@"";
    
    if(url.fragment)
        ret[@"fragment"]=url.fragment;
    else
        ret[@"fragment"]=@"/";

    ret[@"query"]= url.uq_queryDictionary;

    
    if(dict[@"hideNavbar"]){
        ret[@"params"] = @{@"hideNavbar":dict[@"hideNavbar"]};
    }

    return ret;
}

- (void) openTargetRouter:(NSDictionary*) dict complete:(XEngineCallBack)completionHandler {
    NSDictionary* directDTO = [JSIOldRouterModule convertRouter2JSIModel:dict];
    [self.directors push:directDTO[@"scheme"] host:directDTO[@"host"] pathname:directDTO[@"pathname"] fragment:directDTO[@"fragment"] query:directDTO[@"query"] params:directDTO[@"params"]];
}
@end

File: modules/x-engine-jsi-legacy/android/jsi-legacy/src/main/java/com/zkty/jsi/JSIOldRouterModule.java
package com.zkty.jsi;

import android.text.TextUtils;
import android.webkit.JavascriptInterface;

import com.alibaba.fastjson.JSONObject;
import com.anthonynsimon.url.URL;
import com.anthonynsimon.url.exceptions.MalformedURLException;

import com.zkty.nativ.core.NativeContext;
import com.zkty.nativ.core.NativeModule;
import com.zkty.nativ.direct.IDirectManager;
import com.zkty.nativ.direct.NativeDirect;
import com.zkty.nativ.jsi.JSIModule;
import com.zkty.nativ.jsi.bridge.CompletionHandler;

import java.util.HashMap;
import java.util.Map;

public class JSIOldRouterModule extends JSIModule {

    private NativeDirect directors;

    @Override
    public String moduleId() {
        return "com.zkty.module.router";
    }

    @Override
    protected void afterAllJSIModuleInited() {

        NativeModule module = NativeContext.sharedInstance().getModuleByProtocol(IDirectManager.class);
        if (module instanceof NativeDirect)
            directors = (NativeDirect) module;
    }


    private JSONObject convertRouter2JSIModel(JSONObject obj) {
        JSONObject ret = new JSONObject();

        ret.put("scheme", obj.getString("type"));
        String host = obj.getString("uri");
        String url_need_check = null;
        if ("microapp".equals(obj.getString("type"))) {
            url_need_check = String.format("http://%s#%s", host, obj.getString("path"));
        } else {
            if (host.contains("#")) {
                url_need_check = String.format("%s%s", host, obj.getString("path"));
            } else {
                url_need_check = String.format("%s#%s", host, obj.getString("path"));

            }
        }
        String standard_url = SPAUrl2StandardUrl(url_need_check);
        try {
            URL url = URL.parse(standard_url);

            ret.put("host", url.getHost());

            if (!TextUtils.isEmpty(url.getPath())) {
                ret.put("pathname", url.getPath());
            }
            if (!TextUtils.isEmpty(url.getFragment())) {
                ret.put("fragment", url.getFragment());
            }

            if (url.getQueryPairs() != null) {
                ret.put("query", url.getQueryPairs());
            }
            if (obj.getBoolean("hideNavbar")) {
                Map<String, Boolean> map = new HashMap<>();
                map.put("hideNavbar", true);
                ret.put("params", map);
            }

        } catch (MalformedURLException e) {
            e.printStackTrace();
        }

        return ret;


    }

    //http://www.baidu.com/index.html#/test?id=1
    private String SPAUrl2StandardUrl(String raw) {
        int questionMark = -1;
        int hashtagMark = -1;

        for (int i = 0; i < raw.length(); i++) {
            char cc = raw.charAt(i);
            if (cc == '?' && hashtagMark == -1) {
                hashtagMark = i;
            }
            // 仅当找到 hashtag 后才再找?, 不然不是 SPA url
            if (hashtagMark != -1 && cc == '#' && questionMark == -1) {
                questionMark = i;
            }

        }
        if (questionMark != -1 && hashtagMark != -1) {
            String sub1 = raw.substring(0, hashtagMark);
            String sub2 = raw.substring(hashtagMark, questionMark);
            String sub3 = raw.substring(questionMark, raw.length() - 1);
            return String.format("%s%s%s", sub1, sub3, sub2);

        }
        return raw;

    }


    @JavascriptInterface
    public void openTargetRouter(JSONObject obj, final CompletionHandler<Object> handler) {
        JSONObject data = convertRouter2JSIModel(obj);

        if (data.get("host") == null) return;
        if (directors != null) {
            directors.push(data.getString("scheme"), data.getString("host"), data.getString("pathname"),
                    data.getString("fragment"),
                    data.get("query") == null ? null : (Map<String, Object>) data.get("query"),
                    data.get("params") == null ? null : (Map<String, Object>) data.get("params"));
        }
        handler.complete();

    }

}
